### ISSUE TITLE
Discrepancy in package interface

### ISSUE THREAD

**Issue by raphael** [dimfeld/httptreemux#46]
Hello, I'm trying to leverage the ability to use the request context to retrieve the path parameters and I'm finding that the way to do it doesn't exactly work the way I would want it to. Sorry it's a bit of a long winded explanation but I can't think of a better way to go about it...

* One may create a tree mux with `New()`. That object is a `Group` so exposes methods to register `httptreemux` handlers. It's also a `http.Handler`.
* One may create a group using `NewGroup`. A group makes it possible to register `httptreemux` handlers.
* One may create a context group with `UsingContext` which makes it possible to register `http` handlers.

One cannot however create a tree mux which would both allow registering `http` handlers and expose a `ServeHTTP` method.  At the end of the day that's the interface I wish to have:
```go
	Mux interface {
		http.Handler
		Handle(method, pattern string, handler http.HandlerFunc)
	}
```
This is what I need to do in order to achieve this:
```go
type mux struct {
	r *httptreemux.TreeMux
	g *httptreemux.ContextGroup
}

func NewMux() Mux {
	r := httptreemux.New()
	r.EscapeAddedRoutes = true
	return &mux{r: r, g: r.UsingContext()}
}

func (m *mux) Handle(method, pattern string, handler http.HandlerFunc) {
	m.g.Handle(method, pattern, handler)
}

func (m *mux) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	m.r.ServeHTTP(w, r)
}
```
Really I wish I could just do `httptreemux.NewWithContext()` or something like that which would return an object that exposes both `Handle` (accepting `http.HandlerFunc`) and `ServeHTTP`.

Hope that makes sense....

**Comment by dimfeld** [comment]
Yeah, that's a good idea. I think a solution that works somewhat like your last example, but internal to the package and using an embedded member for the mux should work. I'll think about the best way to go about this.

**Comment by raphael** [comment]
Great, thank you for your consideration!

**Comment by token419** [comment]
Hey @raphael I'm not sure of your exact request, but it seems similar to mine, where I wished to have the default returned group on a new router, be a context group, so that I could switch to all native http.Handler types, instead of using the ones with different interface signatures now that Context landed in 1.7.

If that's the functionality you are looking for, I was able to achieve this like so:
	router := httptreemux.New()
	root := router.Group.UsingContext()

	root.GET("/:id", func(w http.ResponseWriter, r *http.Request) {
		params := httptreemux.ContextParams(r.Context())
		id := params["id"]
		fmt.Fprintf(w, "GET /%s", id)
	})

	log.Fatal(http.ListenAndServe(":9090", router)

As far as I can tell, the httptreemux.ServeHTTP takes in a standard request/writer signature as well. But I don't use httptreemux.ServeHTTP directly in my code, so I’m not sure if this address your needs fully.

**Comment by dimfeld** [comment]
My understanding is that the request here is just to have a single object that combines `root` and `router`, so you can just pass around that one object, create the routes, and then pass it to `ListenAndServe`.

This month is really busy for me, so I likely won't get to this until March, but rest assured, I haven't forgotten about it :)

**Comment by raphael** [comment]
Yes exactly, right now I need to keep references to two different objects - one to add the routes and one to serve - and I wish I could just do both on the same object.

**Comment by dimfeld** [comment]
The `NewContextMux` function in 86f7c217d9043ebc6adfd8e2ed04a0bb1e1db651 should take care of this for you. Released as `v3.8.0`

**Comment by raphael** [comment]
Perfect, thank you!
